Bluetooth: controller: Add HCI_LE_Set_Connectionless_CTE_Transmit_Parameters command and its handling

ppryga-nordic opened the pull request:
Add implementation of HCI_LE_Set_Connectionless_CTE_Transmit_Parameters command in HCI.
Add implementation of command handling in controller.

There is one thing I'd like to point at here. DF periodic advertising configuration is stored in `lll_df_adv_cfg` structures.
There were two options to connect it with `lll_adv_sync`, as direct member (by composition) or as pointer to struct (by aggregation).
I decided to go with pointer because I wanted to achieve good decoupling of DF and other parts of controller.
Pros or that approach, there is no need to add lll_df.h header file to all source files where lll_adv.h is included.
It would be added to files where it would actually be used. So contents of the type and any further changes related with it should be visible in limited number of source files (actually only ull_df.c and lll_df.c).

Drawback of the solution are:
- usage of additional few bytes of memory for: pointer in `lll_adv_sync`, and pointer to store free memory pool `df_adv_cfg_free`.
- additional handling related with acquire and release of memory for `lll_df_adv_cfg`.

I hope this is a good way to implement the feature and will be accepted by codeowners.

ppryga-nordic committed:
Bluetooth: controller: Add set connectionless CTE TX parameters

Add implementation of HCI_LE_Set_Connectionless_CTE_-
Transmit_Parameters command in HCI.
Add implementation of command handling in controller.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

cvinayak reviewed (dismissed):
Add unit tests like in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/29947

Else the code will not be covered in the PR CI.

ppryga commented:
Tests are added. 
I've added `HAS_HW_NRF_RADIO_BLE_DF` to nrf52_bsim to make possible to run tests.
Also I had to remove `ull_df_adv_cfg_release`, because it was unused and caused sanity check build to fail.
It is temporary, function will be restored when df_adv_cfg cleanup is added.

aescolar commented on boards/posix/nrf52_bsim/Kconfig.board:
```diff
@@ -24,6 +24,8 @@ config BOARD_NRF52_BSIM
 	select HAS_HW_NRF_POWER
 	select HAS_NRFX
+	# Indicate Direction Finding Extension is supported
+	select HAS_HW_NRF_RADIO_BLE_DF
```
This does not seem like the right thing to do.
There is no actual support for direction finding in the nrf52_bsim (52832 HW models).
It seems you want this only to be able to compile the DF unit tests which do not require the HW anyway. So maybe it should be possible to enable the needed BT_CTLR_DF_* options directly without setting this option? (@cvinayak any thoughts?)

ppryga replied:
Correct I've done it just to run tests.
As far as I know, I have to extend nrf52_bsim to handle DF there. 
If that is correct, then I need to wanted to do in "on demand" approach. When implement some DF functionality and need to test it then add missing parts to nrf52_bsim. 
There are two questions:
- should I leave it as it is now or change board in yaml to real DK?
- is nrf52_bsim designed to be nrf52832 or can it handle other nRF SOC also?

aescolar replied:
>As far as I know, I have to extend nrf52_bsim to handle DF there.

Maybe.. Note that the HW models won't be really able to support direction finding. You could build some empty stubbing of the registers in HW, or maybe a phony HW model which pretends to do something, but really does nothing (I have not looked into what the direction finding HW in the 833 does). But you won't be able to support simulated direction finding, as the physical layer simulation does not support it.

>should I leave it as it is now or change board in yaml to real DK?

Do you mean change the test so it runs only on a real nrf52833?

>is nrf52_bsim designed to be nrf52832 or can it handle other nRF SOC also?

The nrf52_bsim uses the nrf52 HW models, which as today target a 52832 (I needed to chose something whenever there was differences, so I targeted that one). I actually haven't looked in detail into what is different in a 52833.
A relevant thing: the nrf52_bsim does not have any configuration to select modeling one or another IC. If you guys would want to start having several SOCs options in the nrf52_bsim we'd need to add some config options to select which.

ppryga replied:
> Maybe.. Note that the HW models won't be really able to support direction finding. You could build some empty stubbing of the registers in HW, or maybe a phony HW model which pretends to do something, but really does nothing (I have not looked into what the direction finding HW in the 833 does). But you won't be able to support simulated direction finding, as the physical layer simulation does not support it.

Yes, I know that. In my humble opinion, simulation of doing something on air is enough to implement tests. To validate code, there is a need to have expected HW behavior only. So phony stubs should be enough.

> Do you mean change the test so it runs only on a real nrf52833?

Yes, that is what I meant. 

> The nrf52_bsim uses the nrf52 HW models, which as today target a 52832 (I needed to chose something whenever there was differences, so I targeted that one). I actually haven't looked in detail into what is different in a 52833.
A relevant thing: the nrf52_bsim does not have any configuration to select modeling one or another IC. If you guys would want to start having several SOCs options in the nrf52_bsim we'd need to add some config options to select which.

The thing is, DF is not supported by all nRF SoC. To make it possible to tests controller for DF, the missing implementation in nrf52_board should be allowed. Or we have to test controller with use  of real hardware. 
I'm not sure if approach to force build system to enable some DF Kconfigs is desired. 
What do you thing guys?

cvinayak replied:
> It seems you want this only to be able to compile the DF unit tests which do not require the HW anyway. So maybe it should be possible to enable the needed BT_CTLR_DF_* options directly without setting this option? 

As the real h/w will not be used in simulation, introduce HAL interface for logical functionalities (which in turn use the nRFx interfaces on DK). For the unit tests, these stubs be defined in the tests.

> should I leave it as it is now or change board in yaml to real DK?
No, unit tests be in nrf52_bsim, and stubs handle the test case requirements.

> is nrf52_bsim designed to be nrf52832 or can it handle other nRF SOC also?
Currently nrf52_bsim is restricted to nRF52832 feature set. If there is requirement in the future, we can migrate it to support latest new h/w features. But we will only keep single simulation target (it will be expensive to maintain multiple simulation targets).

cvinayak committed:
Bluetooth: controller: Clean up direction finding Kconfig

Simplify selection of direction finding related Kconfig.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: ull_df: Fix compilation warnings

Remove unused functions to fix compilation warninigs.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: bluetooth: Add unit tests for set connectionless CTE parameters

Add unit tests for Direction Finding set of connectionless TX
CTE parameters. Enable DF in nrf52_bsim board.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -164,7 +158,8 @@ uint8_t ll_df_set_cl_cte_tx_params(uint8_t adv_handle, uint8_t cte_len,
 	     cte_type == BT_HCI_LE_AOD_CTE_2US) &&
 	    (num_ant_ids < LLL_DF_MIN_ANT_PATTERN_LEN ||
-	     num_ant_ids > CONFIG_BT_CTLR_DF_MAX_ANT_SW_PATTERN_LEN)) {
+	     num_ant_ids > CONFIG_BT_CTLR_DF_MAX_ANT_SW_PATTERN_LEN ||
+	     !ant_ids)) {
```
Change moved to separate commit.

ppryga commented:
@cvinayak Thank you very much for providing changes related with nrf52_bsim. 
I've split commit that adds tests to move changes in ull_df.c file.

cvinayak commented:
The changes in this PR look good to me, I let others to review

aescolar reviewed (commented):
I'm very fine with the solution to the kconfig dependency.

One note: Beware that you have placed the test in the bsim_bt/ folder. So far, there, there are only tests that are compiled and run outside of sanitycheck. It is not much of a problem to place other things there though, only that the README.txt description is not correct :)

cvinayak commented:
> Beware that you have placed the test in the bsim_bt/ folder. So far, there, there are only tests that are compiled and run outside of sanitycheck.

Ok. My oversight, let me check how other controller focused tests are placed, may be the tests/bluetooth/df was originally the correct location?

ppryga commented:
DF tests were moved to previous location `tests/bluetooth/df`.

aescolar reviewed (approved):
+1 No complaints from my side. But others should approve

ppryga commented:
I've fixed an error in ll_df_set_cl_cte_tx_params.

aescolar commented:
One thing, (apart from rebasing), please squash the "reverting .." commit and so forth :)
(You can remove the DNM after)

joerchan commented on tests/bluetooth/df/Kconfig:
```diff
@@ -0,0 +1,9 @@
+
+config BT_CTLR_DF_SUPPORT
+	bool
+	default y
```
I don't understand why this is needed, isn't it enough to enable it in the prj.conf file?

ppryga replied:
Without it BT_CTLR_DF_SUPPORT will be enabled when nrf52833 board is selected.
When compile UT  the nrf52_bsim board is selected, so no DF code is compiled in controller.
Also this is hidden entry, it will generate warnings during build when selected in prj.conf (correct me if I'm wrong).

ppryga commented:
Commits:
- `boards: nrf52_bsim: revert selecting HAS_HW_NRF_RADIO_BLE_DF`
- `tests: Bluetooth: df: Enable support for DF in controller`
were squashed into commit `tests: bluetooth: Add unit tests for set connectionless CTE parameters`
Commit `Bluetooth: controller: Clean up direction finding Kconfig` was moved before commit `tests: bluetooth: Add unit tests for set connectionless CTE parameters`

ppryga commented:
@jhedberg @joerchan @Vudentz @carlescufi @cvinayak could you do a review?
